test/hardhat/RouterUpgrade.test.ts

```diff
@@ -0,0 +1,355 @@
+    await routerImplementation.waitForDeployment();
+
+    // Deploy UUPS proxy for Router using the implementation address and initialize data
+    const UUPSProxy = new ethers.ContractFactory(UUPSProxy__factory.abi, UUPSProxy__factory.bytecode, owner);
+    const routerProxy = await UUPSProxy.deploy(
+      await routerImplementation.getAddress(),
+      Router.interface.encodeFunctionData("initialize", [
+        ownerAddr,
+        await bn254SigScheme.getAddress(),
+        await bn254SigScheme.getAddress(),
+        VERIFICATION_FEE_BPS,
+      ]),
+    );
+    await routerProxy.waitForDeployment();
+
+    // Attach Router interface to proxy address
+    const routerAttached = Router__factory.connect(await routerProxy.getAddress(), owner);
+    router = routerAttached;
+
+    // Router contract configuration
+    await router.connect(owner).permitDestinationChainId(DST_CHAIN_ID);
+    await router.connect(owner).setTokenMapping(DST_CHAIN_ID, await dstToken.getAddress(), await srcToken.getAddress());
+  });
+
+  describe("scheduleUpgrade", () => {
+    it("should schedule an upgrade with valid params (good path)", async () => {
+      const version = await router.getVersion();
+      expect(version).to.equal("1.0.0");
+
+      const newImplementation: Router = await new MockRouterV2__factory(owner).deploy();
+      await newImplementation.waitForDeployment();
+      const newImplAddress = await newImplementation.getAddress();
+      const latestBlock = await ethers.provider.getBlock("latest");
+      const upgradeTime = latestBlock ? latestBlock.timestamp + 3600 : 0; // 1 hour in the future
+      await expect(router.connect(owner).scheduleUpgrade(newImplAddress, "0x", upgradeTime))
+        .to.emit(router, "UpgradeScheduled")
+        .withArgs(newImplAddress, upgradeTime);
+    });
+
+    it("should revert if new implementation address is zero (bad path)", async () => {
+      const latestBlock = await ethers.provider.getBlock("latest");
+      const upgradeTime = latestBlock ? latestBlock.timestamp + 3600 : 0; // 1 hour in the future
+      await expect(router.connect(owner).scheduleUpgrade(ZeroAddress, "0x", upgradeTime)).to.be.revertedWithCustomError(
+        router,
+        "ZeroAddress()",
+      );
+    });
+
+    it("should revert if upgrade time is not in the future (bad path)", async () => {
+      const newImplementation: Router = await new MockRouterV2__factory(owner).deploy();
+      await newImplementation.waitForDeployment();
+      const newImplAddress = await newImplementation.getAddress();
+      const latestBlock = await ethers.provider.getBlock("latest");
+      const upgradeTime = latestBlock ? latestBlock.timestamp - 10 : 0; // 10 seconds in the past
+      await expect(
+        router.connect(owner).scheduleUpgrade(newImplAddress, "0x", upgradeTime),
+      ).to.be.revertedWithCustomError(router, "UpgradeTimeMustBeInTheFuture()");
+    });
+
+    it("should revert if called by non-admin (bad path)", async () => {
```

Comment:
can remove this as long as threshold sig is good

Reply:
It's been replaced with a test for `should not revert if called by non-admin (good path)`.